### PR TITLE
Allow creating namespaced domelements

### LINKED ISSUE #306: Support creation of namespaced elements
This relates to https://github.com/ckeditor/ckeditor5/issues/1842.

Currently it's not possible to use createElement to create working namespaced elements like SVGs. 

PR is on its way.

### PR DESCRIPTION
**Pull request by Sebobo**
Fix: Create valid namespaced elements. Closes #306.

---

### Additional information

* Relates to https://github.com/ckeditor/ckeditor5/issues/1842

### CODE REVIEW

**Comment by mlewand** [comment]
See ckeditor5#2088.

File: src/dom/createelement.js [hunk 1]
```diff
@@ -26,7 +26,8 @@ import { isString } from 'lodash-es';
  * @returns {Element} Created element.
  */
 export default function createElement( doc, name, attributes = {}, children = [] ) {
-	const element = doc.createElement( name );
+	const namespace = attributes instanceof Object ? attributes.xmlns : null;
+	const element = namespace ? doc.createElementNS( namespace, name ) : doc.createElement( name );
```
Comment (Reinmar) [inline src/dom/createelement.js]: I think that you don't need to check whether `attributes` is an object. You can simply assume it is (which happens later in this function anyway).

File: tests/dom/createelement.js [hunk 2]
```diff
@@ -23,6 +23,15 @@ describe( 'createElement', () => {
 		expect( p.getAttribute( 'class' ) ).to.equal( 'foo' );
 	} );
 
+	it( 'should create element with namespace', () => {
+		const namespace = 'http://www.w3.org/2000/svg';
+		const svg = createElement( document, 'svg', { xmlns: namespace } );
+
+		expect( svg.tagName.toLowerCase() ).to.equal( 'svg' );
+		expect( svg.getAttribute( 'xmlns' ) ).to.equal( namespace );
+		expect( svg.createSVGRect instanceof Function ).to.equal( true );
```
Comment (Reinmar) [inline tests/dom/createelement.js]: I think this should work:

```suggestion
		expect( svg.createSVGRect ).to.be.a( 'function' );
```

**Comment by Reinmar** [comment]
Thanks for the PR. I found two small issues and I think that's it.

File: src/dom/createelement.js (hunk 1 above)
Comment (Reinmar) [reply src/dom/createelement.js]: > You can simply assume it is (which happens later in this function anyway).

Actually, I thought there's `for-of` loop next, which would verify that `attributes` is an object. But it's `for-in` so it'll work fine with `undefined` and `null`. So, if other tests or the reality requires that you handle `attributes` being `null` or `undefined, then this should be fine:

```js
const namespace = attributes && attributes.xmlns;
```

Comment (Sebobo) [reply src/dom/createelement.js]: I actually had the problem that in some tests attributes was null and my code failed.

File: tests/dom/createelement.js (hunk 2 above)
Comment (Sebobo) [reply tests/dom/createelement.js]: Cool, thanks for the hint!

File: src/dom/createelement.js (hunk 1 above)
Comment (Sebobo) [reply src/dom/createelement.js]: So based on that, should I still change something or leave it like that?

Comment (Reinmar) [reply src/dom/createelement.js]: I think you could simplify this to:

```js
const namespace = attributes && attributes.xmlns;
```

I'll try this on `master` after merging your PR.

**Comment by Reinmar** [comment]
Thanks again! ❤

**Pull request merged**